layout: Skip layout traversals that obviously won't do anything.

pcwalton opened the pull request:
This reduces CPU usage when mousing over simple pages (example.com). More complex pages (Wikipedia) still reflow a lot due to other bugs.

Additionally, this change causes Servo to stop painting the results of hit test queries. This is also a win for CPU usage.

This significantly improves #9999, though there's more that can be done. I'll leave it open in case @paulrouget thinks this PR isn't enough.

r? @mbrubeck

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10021)

<!-- Reviewable:end -->

pcwalton committed:
layout: Don't rebuild display lists at all unless restyling tells us
some object needs to be repainted.

Reduces CPU usage when mousing over simple documents. (More complex
documents tend to trigger unnecessary reflow bugs and so still have high
CPU.)

Part of #9999.

pcwalton committed:
layout: Don't start the reflow pass unless some node needs to be
reflowed.

Reduces CPU usage when mousing over simple pages.

Part of #9999.

pcwalton committed:
script: Don't paint layouts induced by hit test queries.

Reduces CPU usage when mousing over simple pages.

Part of #9999.

pcwalton committed:
layout: Minor whitespace cleanup.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout and script code, but no tests are modified. Please consider adding a test!

ecoal95 commented on components/layout/layout_thread.rs:
```diff
@@ -869,49 +870,58 @@ impl LayoutThread {
+                let display_list = if flow::base(&**layout_root).restyle_damage
+                                                                .contains(REPAINT) ||
+                        rw_data.display_list.is_none() {
+                    println!("rebuilding display list");
```
nit: Probably `debug!` is wanted here.

mbrubeck commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/layout/layout_thread.rs, line 873 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10021#-KCwoZl_Jnpg3V2hnHOe:-KCwoZl_Jnpg3V2hnHOf:-1580885930) ([raw file](https://github.com/servo/servo/blob/01ca1153b2884a5ded369eb94f333b37f7a874e7/components/layout/layout_thread.rs#L873)):*
With this change, I got a layout thread panic at 'Tried to hit test with no display list' when mousing over the window before a page loaded.

Maybe if there is no display list, we should build it unconditionally?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10021#-:-KCwouSd5z_YA6mJXsGC:1458266033)*
<!-- Sent from Reviewable.io -->

pcwalton commented:
Addressed review comments. r? @mbrubeck

Ms2ger commented on components/layout/traversal.rs:
```diff
@@ -236,6 +236,6 @@ impl<'a> BuildDisplayList<'a> {
 
     #[inline]
     fn should_process(&self) -> bool {
```
Remove?

paulrouget commented:
I think this is breaking browserhtml.

The progressbar is not redrawn.

paulrouget commented:
Also - PNG images show up late (the window controls in browserhtml).

paulrouget commented:
Would a test case help?

pcwalton commented:
Yeah, a test case would be useful. This sounds like another case in which we didn't reflow properly in script and counted on the hit test reflow to draw.

mbrubeck commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10021)*
<!-- Sent from Reviewable.io -->

mbrubeck commented:
> PNG images show up late (the window controls in browserhtml).

I think this is because an image load increments the `pending_reflow_count`, but then a hit test test will cause a call to `Window::force_reflow` which resets the counter (cancels the pending reflow).  This is now incorrect because the forced reflow does not repaint, while the pending reflow would have.

pcwalton committed:
layout: Make transform changes require incremental repaint.

The progress bar in browser.html was only accidentally working before,
so actually making use of `REPAINT` makes it stop working without this
change.

pcwalton commented:
Fixed both bugs.

r? @mbrubeck

mbrubeck commented:
@bors-servo r+

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10021#-:-KD0RvyqXMyzn_iowE0k:1236647940)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit df27547 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck df275470cb75ae76bdf62bf8358b8624169a2702 -->

bors-servo commented:
:hourglass: Testing commit df27547 with merge 4f5999c...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2178)

mbrubeck commented:
There are unexpected results in both test-css and test-wpt, and they are consistent across Linux and Mac:

```
  ▶ PASS [expected FAIL] /css-transforms-1_dev/html/perspective-translateZ-negative.htm

  ▶ FAIL [expected PASS] /css21_dev/html4/z-index-dynamic-001.htm
  └   → /css21_dev/html4/z-index-dynamic-001.htm 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7
/css21_dev/html4/reference/ref-filled-green-100px-square.htm 82d94d2027e216f7a2d31fcc53afb5e048516cef
Testing 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7 == 82d94d2027e216f7a2d31fcc53afb5e048516cef

  ▶ FAIL [expected PASS] /_mozilla/css/list_item_overflow.html
  └   → /_mozilla/css/list_item_overflow.html 916b152431c176d3c9b28afb86be616d6168f646
/_mozilla/css/list_item_overflow_ref.html a8fc81d3fc0f23056c2e6a616cba6e4e684a823f
Testing 916b152431c176d3c9b28afb86be616d6168f646 == a8fc81d3fc0f23056c2e6a616cba6e4e684a823f

  ▶ FAIL [expected PASS] /_mozilla/css/position_fixed_overflow_a.html
  └   → /_mozilla/css/position_fixed_overflow_a.html 6f87cc04be15cb35de8a96f996443204a0ea75bf
/_mozilla/css/position_fixed_overflow_b.html 53765186b05853a4705eded01c26237bd847a93c
Testing 6f87cc04be15cb35de8a96f996443204a0ea75bf == 53765186b05853a4705eded01c26237bd847a93c
```

mbrubeck commented:
The `list_item_overflow` and `position_fixed_overflow` failures are identical to the intermittent failures of these tests in #9012 and #9278. So this has probably exposed another existing incremental layout bug, in a way that makes it fail consistently rather than intermittently.

pcwalton commented:
Hmm. Would you rather I try to fix the bug or mark them failing and update those issues?

mbrubeck commented:
> Would you rather I try to fix the bug or mark them failing and update those issues?

Opening the test in Servo locally, it's definitely rendering it wrong consistently.  It's not just temporarily wrong either; it stays wrong unless I force a reflow by resizing the window.  I'd be concerned this is causing other, possibly real-world regressions.  So we should fix the bug or at least figure out more.

pcwalton commented:
So the culprit here is https://github.com/servo/servo/pull/8306. In particular, the problem is that the `position: fixed` elements try to union their kids' overflow regions, but they haven't been calculated yet, so they're zero. This is why `store_overflow` was bottom up.

@glennw Did you have a solution in mind for this? I'm actually confused as to how https://github.com/servo/servo/pull/8306 was a fix in the first place…

glennw commented:
@pcwalton I am confused, looking at that commit - I don't see how that could ever have worked either, sorry! Perhaps it was meant to be done in the DL building pass, since that was bottom up when that commit was made? We really need to somehow get better test coverage for overflow region calculation...

mbrubeck commented:
> I don't see how that could ever have worked either, sorry!

It worked because we were doing extra reflows, so the correct values got propagated when the traversal ran for a second time.

To fix this, can we just make the traversal call `compute_absolute_position` on the way down, and then `store_overflow` on the way back up?

pcwalton commented:
@mbrubeck No, because we build display lists during the top down part of the traversal and overflow has to be known by then.

I'm working on moving overflow calculation back to block size assignment.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10064) made this pull request unmergeable. Please resolve the merge conflicts.

pcwalton committed:
layout: Repaint when `z-index` is modified.

pcwalton commented:
r? @mbrubeck (added second commit to fix a test failure)

mbrubeck commented:
@bors-servo r+

---

Reviewed 4 of 4 files at r4, 1 of 2 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10021#-:-KDPPziOrPVeqpMpgtjo:1236647940)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 7a131ba has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 7a131ba104fe5500ab4bd46ae9c3e35ffba5054d -->

bors-servo commented:
:hourglass: Testing commit 7a131ba with merge e31d475...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/2087)

pcwalton commented:
@larsbergstrom Do you have an idea as to what might be going on with

pcwalton commented:
@larsbergstrom ...that failure?

mbrubeck commented:
It's failing to find `arm-linux-androideabi-gcc`. Looks like a configuration problem on the "servo-linux-cross1" build machine.

larsbergstrom commented:
@pcwalton Sorry, we are doing upgrades to the infra. I should have turned off r+ privs entirely instead of just posting on #servo - will do in the future! I'll retry once the machines are online.

edunham commented:
@bors-servo retry

My bad on the earlier fail; I added a slave to Buildbot before double checking that it could complete a test successfully

bors-servo commented:
:hourglass: Testing commit 7a131ba with merge 9b438ee...

edunham commented:
@bors-servo retry

edunham commented:
Buildbot may fail here due to variable expansion failure fixed in https://github.com/servo/saltfs/pull/261 .

bors-servo commented:
:bomb: Failed to start rebuilding: `Unknown error`

bors-servo commented:
:hourglass: Testing commit 7a131ba with merge 9813d11...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2091), [gonk](http://build.servo.org/builders/gonk/builds/2152), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2129), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2412), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/821), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2178), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2236), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.811)
